torch/csrc/autograd/input_buffer.cpp

```diff
@@ -20,55 +20,6 @@
 namespace torch::autograd {
 
 namespace {
-// look what you made me do >.<
-// Divergent paths for per-Impl stream recording that leak implementation
-// details of the impls should not be needed here.
-// See https://github.com/pytorch/pytorch/issues/60306
-// TODO: clean this up when https://github.com/pytorch/pytorch/issues/60306 is
-// improved
-void record_stream_any_impl(Variable& var, c10::Stream& stream) {
-  // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
-  const auto guard = c10::impl::VirtualGuardImpl(device_of(var).value().type());
-
-  if (C10_UNLIKELY(at::isBatchedTensor(var))) {
-    auto* impl = at::maybeGetBatchedImpl(var);
-    if (impl) {
-      guard.recordDataPtrOnStream(impl->value().storage().data_ptr(), stream);
-    } else {
-      TORCH_INTERNAL_ASSERT(false, "Expected batched tensor");
-    }
-  } else {
-    switch (var.layout()) {
-      case c10::kSparseCsr:
-      case c10::kSparseCsc:
-      case c10::kSparseBsr:
-      case c10::kSparseBsc: {
-        auto* impl = at::sparse_csr::get_sparse_csr_impl(var);
-        guard.recordDataPtrOnStream(
-            impl->values().storage().data_ptr(), stream);
-        guard.recordDataPtrOnStream(
-            impl->compressed_indices().storage().data_ptr(), stream);
-        guard.recordDataPtrOnStream(
-            impl->plain_indices().storage().data_ptr(), stream);
-        break;
-      }
-      case c10::kSparse: {
-        auto* impl = at::sparse::get_sparse_impl(var);
-        guard.recordDataPtrOnStream(
-            impl->values().storage().data_ptr(), stream);
-        guard.recordDataPtrOnStream(
-            impl->indices().storage().data_ptr(), stream);
-        break;
-      }
-      case c10::kStrided:
-        guard.recordDataPtrOnStream(var.storage().data_ptr(), stream);
-        break;
-      default:
-        TORCH_INTERNAL_ASSERT(
-            false, "Unknown layout in record_stream_any_impl");
-    }
-  }
-}
```

Comment:
Code movement

Reply:
To make it possible to use in engine.cpp (related discussion https://github.com/pytorch/pytorch/pull/151079#discussion_r2079914220)